lightning/src/ln/channelmanager.rs

```diff
@@ -3522,22 +3535,27 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> MessageSend
 	}
 }
 
-impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> EventsProvider for ChannelManager<Signer, M, T, K, F, L>
-	where M::Target: chain::Watch<Signer>,
-        T::Target: BroadcasterInterface,
-        K::Target: KeysInterface<Signer = Signer>,
-        F::Target: FeeEstimator,
-				L::Target: Logger,
+impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref, E> EventsProvider<E> for ChannelManager<Signer, M, T, K, F, L>
+where
+	M::Target: chain::Watch<Signer>,
+	T::Target: BroadcasterInterface,
+	K::Target: KeysInterface<Signer = Signer>,
+	F::Target: FeeEstimator,
+	L::Target: Logger,
 {
-	fn get_and_clear_pending_events(&self) -> Vec<Event> {
+	fn process_pending_events<H: Deref>(&self, handler: H) -> Result<(), E>
+	where H::Target: EventHandler<E> {
+		let _persistence_guard = PersistenceNotifierGuard::new(&self.total_consistency_lock, &self.persistence_notifier);
 		//TODO: This behavior should be documented. It's non-intuitive that we query
 		// ChannelMonitors when clearing other events.
 		self.process_pending_monitor_events();
 
-		let mut ret = Vec::new();
 		let mut pending_events = self.pending_events.lock().unwrap();
-		mem::swap(&mut ret, &mut *pending_events);
-		ret
+		for event in pending_events.drain(..) {
+			handler.handle_event(event)?;
```

Comment:
The code as written here will pass back the error of the first encountered error and stop processing events. However, any remaining events are lost because of the `drain`. So this leads to some other questions:
- Should event processing continue if handling one event fails?
- Can a user reasonably recover from an error while handling an event?
- Should unprocessed events remain pending for the next call?

The answer to these will determine what the interface will ultimately look like.

Reply:
Right, if we take an error from users we definitely shouldn't just drop events, we should keep feeding them to users. That said, I'm a little dubious that we should let users provide Errors either. We shouldn't take an error unless we can do something reasonable with it - maybe try the event again later or something, but short of that I'm not sure what an error even *means*.

Reply:
Yeah, I'm starting to think not allowing errors is the best approach. The user can always write an `EventHandler` to do whatever they want (including collecting errors or remaining events) and act upon it after `process_pending_events` returns. But, depending on how implemented, it could leave them in a state where `ChannelManager` is persisted without those events but the user hasn't "fully" handled them either. May just need to spell it out in the documentation.

That said, if they are using `BackgroundProcessor` there would be no way for them to examine the handler after processing.

Reply:
Right, I guess the only case where errors make sense is some kind of transitory issue that we know will resolve very soon, allowing us to handle the error momentarily while holding off persisting the ChannelManager. That said, I think you could just block in the handler in that case, as long as we only hold an event-specific lock?

Maybe we can have a separate lock allowing us to block any other event handlers from entering while still allowing adding new events from other things going on?

Reply:
> Right, I guess the only case where errors make sense is some kind of transitory issue that we know will resolve very soon, allowing us to handle the error momentarily while holding off persisting the ChannelManager. That said, I think you could just block in the handler in that case, as long as we only hold an event-specific lock?
> 

Will update this to remove the return type from `handle_event`.

> Maybe we can have a separate lock allowing us to block any other event handlers from entering while still allowing adding new events from other things going on?

`pending_events` is already behind a `Mutex`. But I guess if we want to allow other event producers to add more events, we could immediately drop the lock after either draining fully into a `Vec` or swapping as before.

Reply:
> pending_events is already behind a Mutex. But I guess if we want to allow other event producers to add more events, we could immediately drop the lock after either draining fully into a Vec or swapping as before.

Right, I guess I was thinking that if a user were handling events in parallel it would be a bug. Now that I think about it more maybe its fine, as long as that's what the user wants.

Reply:
FWIW, swapping retains the same behavior as `get_and_clear_pending_events` since there the lock was held for the duration of the method and then the events were handled by the user. I updated the code accordingly.